### DISCUSSION TITLE
Attempting to build Chat example - no "secio" feature in "libp2p"

### DISCUSSION THREAD

**Discussion by brandonmarzolf** [libp2p/rust-libp2p#3568] [Q&A] [answered]
I'm attempting to build the chat example to learn the libp2p library.

I'm on Ubuntu, I installed the protobuf compiler, added all the features, including "secio" to the feature flags list in the Cargo.toml file, and my attempt to compile has me stuck on this error:

> error: failed to select a version for `libp2p`.
>     ... required by package `chat v0.1.0 (...src/examples/chat)`
> versions that meet the requirements `^0.51.0` are: 0.51.0
> 
> the package `chat` depends on `libp2p`, with features: `secio` but `libp2p` does not have these features.
> 
> 
> failed to select a version for `libp2p` which could resolve this conflict

From what investigation I've been able to do, being new to the library, it does indeed look like there is no "secio" feature. That has me wondering if the example is broken.

What can I do to resolve this? Is this a broken example, or is something else wrong?

**Comment by thomaseizinger** [comment, marked as answer]
Secio has been removed 2.5 years ago and I cannot find any references to it in the repository. Where does it say you should add the secio feature?

**Reply by brandonmarzolf** [reply]
In the error message, it is just telling me that it's not provided by the library. I posted the entire error message in the above question.

**Reply by brandonmarzolf** [reply]
What I think would be exceedingly helpful with the examples is if each one was project that could be individually compiled. What I'm struggling with is figuring out all the dependencies in order to just get the chat example to compile.

For example, if I remove all the features I have listed, I get many different compile errors which appear to have nothing to do with features:

- could not find `main` in `async_std`
- cannot find derive macro `NetworkBehaviour` in this scope
- cannot find attribute `behaviour` in this scope
- failed to resolve: could not find `async_io` in `mdns`
- cannot find function `development_transport` in crate `libp2p`
- the trait bound `MyBehaviour: NetworkBehaviour` is not satisfied
- `main` function is not allowed to be `async

There are many more errors listed, I just picked out the first and last few. 

Again, it would be exceedingly helpful if the examples were projects that could just be built, instead of leaving it up to users to turn a given example into a project and then try to figure out what all the dependencies are and what is causing the compile errors.

**Reply by thomaseizinger** [reply]
> In the error message, it is just telling me that it's not provided by the library. I posted the entire error message in the above question.

What I don't understand is where you even got the idea from to specify the `secio` feature? Are you following an outdated, external blogpost or something? Within our workspace, I can't find any mention of the `secio` feature apart from changelog entries.

> What I think would be exceedingly helpful with the examples is if each one was project that could be individually compiled. What I'm struggling with is figuring out all the dependencies in order to just get the chat example to compile.

You mean like this: https://github.com/libp2p/rust-libp2p/pull/3509? :)

**Reply by brandonmarzolf** [reply]
> What I don't understand is where you even got the idea from to specify the `secio` feature? Are you following an outdated, external blogpost or something? Within our workspace, I can't find any mention of the `secio` feature apart from changelog entries.

I didn't "chose" to add the feature. I just got the error trying to build. So I added the feature and that didn't resolve anything.

> You mean like this: #3509? :)

They are all in one place, but they don't build individually as they are dependent on the libp2p in the project itself. This is an understandable design choice, however when I switch this out to the latest version released on crates.io, which is necessary if I'm to build a standalone project, the build fails because of version conflicts:

> error: failed to select a version for `subtle`.
>     ... required by package `digest v0.10.6`
>     ... which satisfies dependency `digest = "^0.10.3"` (locked to 0.10.6) of package `blake2 v0.10.6`
>     ... which satisfies dependency `blake2 = "^0.10"` of package `snow v0.9.1`
>     ... which satisfies dependency `snow = "^0.9.0"` of package `libp2p-noise v0.42.0`
>     ... which satisfies dependency `libp2p-noise = "^0.42.0"` of package `libp2p v0.51.0`
>     ... which satisfies dependency `libp2p = "^0.51.0"` (locked to 0.51.0) of package `chat v0.1.0 (/workspaces/INR_DEV/platform/DRIVE_Sync/src/examples/chat)`
> versions that meet the requirements `=2.4` are: 2.4.1, 2.4.0
> 
> all possible versions conflict with previously selected packages.
> 
>   previously selected package `subtle v2.5.0`
>     ... which satisfies dependency `subtle = "^2.2.1"` (locked to 2.5.0) of package `curve25519-dalek v3.2.0`
>     ... which satisfies dependency `curve25519-dalek = "^3"` (locked to 3.2.0) of package `ed25519-dalek v1.0.1`
>     ... which satisfies dependency `ed25519-dalek = "^1.0.1"` (locked to 1.0.1) of package `libp2p-core v0.39.0`
>     ... which satisfies dependency `libp2p-core = "^0.39.0"` (locked to 0.39.0) of package `libp2p v0.51.0`
>     ... which satisfies dependency `libp2p = "^0.51.0"` (locked to 0.51.0) of package `chat v0.1.0 (/workspaces/INR_DEV/platform/DRIVE_Sync/src/examples/chat)`
> 
> failed to select a version for `subtle` which could resolve this conflict

My attempts to get around this error lead to where I am now. Admittedly I've probably gone down the wrong path, but I didn't see any instructions that explained how to get the examples to work on their own.

Perhaps I need to back up and instead ask: How are the examples intended to be built?

**Reply by brandonmarzolf** [reply]
@thomaseizinger Given the last part of my reply, perhaps it is best to close this question, if you agree that I have gone down the wrong path and the error in question is irrelevant? Then I will ask a new question, specific to how to go about building the examples? Does this sound good to you?

**Reply by thomaseizinger** [reply]
> > You mean like this: #3509? :)
> 
> They are all in one place, but they don't build individually as they are dependent on the libp2p in the project itself.

They are built as part of our CI so yes they do build individually. That was one of the main reasons I wanted them separate. They are now copy-pastable.

> when I switch this out to the latest version released on crates.io which is necessary if I'm to build a standalone project

That isn't gonna work, sorry. But that is not specific to our project, this won't reliably work in any Rust project. If you look at examples, I'd advise you to always look at the released Git tag or specify a Git dependency in your project:

```toml
libp2p = { git = "https://github.com/libp2p/rust-libp2p" }
```

> Perhaps I need to back up and instead ask: How are the examples intended to be built?

In terms of selecting features, the easiest is to just specify `full`. Before #3509, we also used that for the examples: https://github.com/libp2p/rust-libp2p/blob/2c10cd838aa0250b05fae5c31a70f3d2a65a48da/Cargo.toml#L183-L185

Now, with #3509 we can finally be more fine-granular and specify the individual features for each example.

In your specific case, you should also be able to copy the feature-list from the `chat` example and depend on the latest released version from `crates.io` (because we haven't made any changes to features since the last release). Let me know if that works.

In regards to your other compile errors with `async_std` etc, you should also be a long way if you look at the `Cargo.toml` of the new examples.

**Reply by thomaseizinger** [reply]
```diff
diff --git a/examples/chat-example/Cargo.toml b/examples/chat-example/Cargo.toml
index bfd2d7e0..06c59736 100644
--- a/examples/chat-example/Cargo.toml
+++ b/examples/chat-example/Cargo.toml
@@ -10,4 +10,4 @@ async-std = { version = "1.12", features = ["attributes"] }
 async-trait = "0.1"
 env_logger = "0.10.0"
 futures = "0.3.26"
-libp2p = { path = "../../libp2p", features = ["async-std", "dns", "gossipsub", "mdns", "mplex", "noise", "macros", "tcp", "websocket", "yamux"] }
+libp2p = { version = "0.51", features = ["async-std", "dns", "gossipsub", "mdns", "mplex", "noise", "macros", "tcp", "websocket", "yamux"] }
```

I just applied the above patch to master and `cargo run --bin chat-example` worked without problems. If you get version conflicts, I'd suggest you nuke your `Cargo.lock` and try again.

**Reply by brandonmarzolf** [reply]
Thank you for all of that.

Since the latest release doesn't have the examples in their own projects, I copied the chat.rs example from the latest release tag, and then I copied the dependencies from the latest release.

This didn't work.

So then, instead of trying to figure out what features to enable, I just enabled "full" as you recommended and that built, and appears to be working.

Thank you for all your help in this!

**Reply by thomaseizinger** [reply]
Glad you got it working. You should be able to replace `full` with the list of features that is now listened in `examples/chat-example/Cargo.toml`. That would likely reduce (initial) compile times a bit.
